core/cp4d_authenticator.go

```diff
@@ -144,24 +151,80 @@ func (authenticator *CloudPakForDataAuthenticator) Authenticate(request *http.Re
 }
 
 // getToken: returns an access token to be used in an Authorization header.
-// Whenever a new token is needed (when a token doesn't yet exist, or the existing token has expired),
-// a new access token is fetched from the token server.
+// Whenever a new token is needed (when a token doesn't yet exist, needs to be refreshed,
+// or the existing token has expired), a new access token is fetched from the token server.
 func (authenticator *CloudPakForDataAuthenticator) getToken() (string, error) {
 	if authenticator.tokenData == nil || !authenticator.tokenData.isTokenValid() {
-		tokenResponse, err := authenticator.requestToken()
+		// synchronously request the token
+		err := authenticator.synchronizedRequestToken()
 		if err != nil {
 			return "", err
 		}
-
-		authenticator.tokenData, err = newCp4dTokenData(tokenResponse)
-		if err != nil {
-			return "", err
+	} else if authenticator.tokenData != nil && authenticator.needsRefresh() {
+		// If refresh needed, kick off a go routine in the background to get a new token
+		ch := make(chan error)
+		go func() {
+			ch <- authenticator.getTokenData()
+		}()
+		select {
+		case err := <-ch:
+			if err != nil {
+				return "", err
+			}
+		default:
 		}
 	}
 
+	// return an error if the access token is not valid or was not fetched
+	if authenticator.tokenData == nil || authenticator.tokenData.AccessToken == "" {
+		return "", fmt.Errorf("Error while trying to get access token")
+	}
+
 	return authenticator.tokenData.AccessToken, nil
 }
 
+// synchronizedRequestToken: synchronously checks if the current token in cache
+// is valid. If token is not valid or does not exist, it will fetch a new token
+// and set the tokenRefreshTime
+func (authenticator *CloudPakForDataAuthenticator) synchronizedRequestToken() error {
+	cp4dMuOne.Lock()
+	defer cp4dMuOne.Unlock()
+	// if cached token is still valid, then just continue to use it
+	if authenticator.tokenData != nil && authenticator.tokenData.isTokenValid() {
+		return nil
+	}
+
+	err := authenticator.getTokenData()
+	if err != nil {
+		return err
+	}
+
+	// If the refresh time isn't set, then set it
+	if authenticator.tokenRefreshTime == 0 && &authenticator.tokenData.ExpiresIn != nil && &authenticator.tokenData.Expiration != nil {
```

Comment:
@mkistler @padamstx - now that I'm looking at this again, should I *always* be setting the correct refresh time when we request a new token? Currently, it will only be set when you request the first token, and when you refresh the token where we push it forward 1 min.. I think this will result in more token requests than we'd like

Reply:
Yes ... you should be updating tokenRefreshTime whenever you get a new token.

Reply:
I realized that as I created the PR 🙂